In [59]:
import pandas as pd
import numpy as np

import seaborn as sns
from matplotlib import pyplot as plt
%matplotlib inline

In [60]:
data = "https://raw.githubusercontent.com/alexeygrigorev/datasets/master/AB_NYC_2019.csv"

In [61]:
!wget $data

--2021-12-26 22:47:34--  https://raw.githubusercontent.com/alexeygrigorev/datasets/master/AB_NYC_2019.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 7077973 (6.8M) [text/plain]
Saving to: 'AB_NYC_2019.csv.1'

     0K .......... .......... .......... .......... ..........  0%  514K 13s
    50K .......... .......... .......... .......... ..........  1%  693K 12s
   100K .......... .......... .......... .......... ..........  2% 1.84M 9s
   150K .......... .......... .......... .......... ..........  2%  399K 11s
   200K .......... .......... .......... .......... ..........  3%  385K 12s
   250K .......... .......... .......... .......... ..........  4% 13.0M 10s
   300K .......... .......... .......... .......... ..........  5%  818K 10s
   350K .......... .......... .......... .......... ...

In [62]:
df = pd.read_csv('AB_NYC_2019.csv')

In [63]:
df.head()

id                                              name  host_id  \
0  2539                Clean & quiet apt home by the park     2787   
1  2595                             Skylit Midtown Castle     2845   
2  3647               THE VILLAGE OF HARLEM....NEW YORK !     4632   
3  3831                   Cozy Entire Floor of Brownstone     4869   
4  5022  Entire Apt: Spacious Studio/Loft by central park     7192   

     host_name neighbourhood_group neighbourhood  latitude  longitude  \
0         John            Brooklyn    Kensington  40.64749  -73.97237   
1     Jennifer           Manhattan       Midtown  40.75362  -73.98377   
2    Elisabeth           Manhattan        Harlem  40.80902  -73.94190   
3  LisaRoxanne            Brooklyn  Clinton Hill  40.68514  -73.95976   
4        Laura           Manhattan   East Harlem  40.79851  -73.94399   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0     Private room    149               1                  9  2018-10-19   
1  Entire home/apt    225               1                 45  2019-05-21   
2     Private room    150               3                  0         NaN   
3  Entire home/apt     89               1                270  2019-07-05   
4  Entire home/apt     80              10                  9  2018-11-19   

   reviews_per_month  calculated_host_listings_count  availability_365  
0               0.21                               6               365  
1               0.38                               2               355  
2                NaN                               1               365  
3               4.64                               1               194  
4               0.10                               1                 0

In [64]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48895 entries, 0 to 48894
Data columns (total 16 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              48895 non-null  int64  
 1   name                            48879 non-null  object 
 2   host_id                         48895 non-null  int64  
 3   host_name                       48874 non-null  object 
 4   neighbourhood_group             48895 non-null  object 
 5   neighbourhood                   48895 non-null  object 
 6   latitude                        48895 non-null  float64
 7   longitude                       48895 non-null  float64
 8   room_type                       48895 non-null  object 
 9   price                           48895 non-null  int64  
 10  minimum_nights                  48895 non-null  int64  
 11  number_of_reviews               48895 non-null  int64  
 12  last_review                     

In [65]:
strings = list(df.dtypes[df.dtypes == 'object'].index)
strings

['name',
 'host_name',
 'neighbourhood_group',
 'neighbourhood',
 'room_type',
 'last_review']

In [66]:
for col in strings:
    df[col] = df[col].str.lower().str.replace(' ', '_')

In [67]:
df.head()

id                                              name  host_id  \
0  2539                clean_&_quiet_apt_home_by_the_park     2787   
1  2595                             skylit_midtown_castle     2845   
2  3647               the_village_of_harlem....new_york_!     4632   
3  3831                   cozy_entire_floor_of_brownstone     4869   
4  5022  entire_apt:_spacious_studio/loft_by_central_park     7192   

     host_name neighbourhood_group neighbourhood  latitude  longitude  \
0         john            brooklyn    kensington  40.64749  -73.97237   
1     jennifer           manhattan       midtown  40.75362  -73.98377   
2    elisabeth           manhattan        harlem  40.80902  -73.94190   
3  lisaroxanne            brooklyn  clinton_hill  40.68514  -73.95976   
4        laura           manhattan   east_harlem  40.79851  -73.94399   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0     private_room    149               1                  9  2018-10-19   
1  entire_home/apt    225               1                 45  2019-05-21   
2     private_room    150               3                  0         NaN   
3  entire_home/apt     89               1                270  2019-07-05   
4  entire_home/apt     80              10                  9  2018-11-19   

   reviews_per_month  calculated_host_listings_count  availability_365  
0               0.21                               6               365  
1               0.38                               2               355  
2                NaN                               1               365  
3               4.64                               1               194  
4               0.10                               1                 0

In [68]:
data = df[['latitude',
'longitude',
'price',
'minimum_nights',
'number_of_reviews',
'reviews_per_month',
'calculated_host_listings_count',
'availability_365']]

Features

In [69]:
data.head()

latitude  longitude  price  minimum_nights  number_of_reviews  \
0  40.64749  -73.97237    149               1                  9   
1  40.75362  -73.98377    225               1                 45   
2  40.80902  -73.94190    150               3                  0   
3  40.68514  -73.95976     89               1                270   
4  40.79851  -73.94399     80              10                  9   

   reviews_per_month  calculated_host_listings_count  availability_365  
0               0.21                               6               365  
1               0.38                               2               355  
2                NaN                               1               365  
3               4.64                               1               194  
4               0.10                               1                 0

Question 1
Find a feature with missing values. How many missing values does it have?

In [70]:
data.isnull().sum()

latitude                              0
longitude                             0
price                                 0
minimum_nights                        0
number_of_reviews                     0
reviews_per_month                 10052
calculated_host_listings_count        0
availability_365                      0
dtype: int64

reviews_per_month has 10052 missing values

Question 2
What's the median (50% percentile) for variable 'minimum_nights'?

In [71]:
data.describe()

latitude     longitude         price  minimum_nights  \
count  48895.000000  48895.000000  48895.000000    48895.000000   
mean      40.728949    -73.952170    152.720687        7.029962   
std        0.054530      0.046157    240.154170       20.510550   
min       40.499790    -74.244420      0.000000        1.000000   
25%       40.690100    -73.983070     69.000000        1.000000   
50%       40.723070    -73.955680    106.000000        3.000000   
75%       40.763115    -73.936275    175.000000        5.000000   
max       40.913060    -73.712990  10000.000000     1250.000000   

       number_of_reviews  reviews_per_month  calculated_host_listings_count  \
count       48895.000000       38843.000000                    48895.000000   
mean           23.274466           1.373221                        7.143982   
std            44.550582           1.680442                       32.952519   
min             0.000000           0.010000                        1.000000   
25%             1.000000           0.190000                        1.000000   
50%             5.000000           0.720000                        1.000000   
75%            24.000000           2.020000                        2.000000   
max           629.000000          58.500000                      327.000000   

       availability_365  
count      48895.000000  
mean         112.781327  
std          131.622289  
min            0.000000  
25%            0.000000  
50%           45.000000  
75%          227.000000  
max          365.000000

The median(50% percentile) of minimum_nights is 3

In [72]:
n = len(data)

n_val = int(n*0.2)
n_test = int(n*0.2)
n_train = n - (n_test + n_val)

In [73]:
n, n_train, n_val, n_test

(48895, 29337, 9779, 9779)

In [74]:
idx = np.arange(n)
np.random.seed(42)
np.random.shuffle(idx)

In [75]:
df_train = data.iloc[idx[: n_train]]
df_val = data.iloc[idx[n_train : n_train + n_val]]
df_test = data.iloc[idx[n_train + n_val :]]

df_train = df_train.reset_index(drop=True)
df_val = df_val.reset_index(drop=True)
df_test = df_test.reset_index(drop=True)

In [76]:
len(df_train), len(df_val), len(df_test)

(29337, 9779, 9779)

In [77]:
y_train =  np.log1p(df_train.price.values)
y_val =  np.log1p(df_val.price.values)
y_test =  np.log1p(df_test.price.values)

In [78]:
del df_train['price']
del df_val['price']
del df_test['price']

Question 3

In [79]:
def train_linear_regression(X, y):
    ones = np.ones(X.shape[0])
    X = np.column_stack([ones, X])

    XTX = X.T.dot(X)
    XTX_inv = np.linalg.inv(XTX)
    w_full = XTX_inv.dot(X.T).dot(y)
    
    return w_full[0], w_full[1:]

In [80]:
def rmse(y, y_pred):
    se = (y - y_pred) ** 2
    mse = se.mean()
    return np.sqrt(mse)

In [81]:
mean = df.reviews_per_month.mean()


X_mean_train = df_train.fillna(mean).values
w0_mean, w_mean = train_linear_regression(X_mean_train, y_train)

In [86]:
X_mean_val = df_val.fillna(mean).values
y_mean_pred_val = w0_mean + X_mean_val.dot(w_mean)
np.round(rmse(y_val, y_mean_pred_val),2)

0.64

In [83]:
X_null_train = df_train.fillna(0).values
w0_null, w_null = train_linear_regression(X_null_train, y_train)

In [85]:
X_null_val = df_val.fillna(0).values
y_null_pred_val = w0_null + X_null_val.dot(w_null)
np.round(rmse(y_val, y_null_pred_val),2)

0.64

**Both are equally good**

Regularization

In [87]:
def train_linear_regression_reg(X, y, r=0.0):
    ones = np.ones(X.shape[0])
    X = np.column_stack([ones, X])

    XTX = X.T.dot(X)
    XTX = XTX + r * np.eye(XTX.shape[0])

    XTX_inv = np.linalg.inv(XTX)
    w_full = XTX_inv.dot(X.T).dot(y)
    
    return w_full[0], w_full[1:]

In [88]:
for r in [0, 0.000001, 0.0001, 0.001, 0.01, 0.1, 1, 5, 10]:
    w0, w = train_linear_regression_reg(X_null_train, y_train, r = r)
    y_null_reg_val = w0 + X_null_val.dot(w)
    rmse_val = np.round(rmse(y_val, y_null_reg_val),2)
    print(r, w0, rmse_val)

0 -419.91265817250303 0.64
1e-06 -419.86271538401513 0.64
0.0001 -414.9764921499109 0.64
0.001 -375.2736523610076 0.64
0.01 -191.7838404491127 0.66
0.1 -32.56256054292286 0.68
1 -3.4992168345626347 0.68
5 -0.7033623163350434 0.68
10 -0.35127676060936247 0.68


The best regulrization parameter is 0

In [97]:
rmse_list = []

for r in [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]:
    
    idx = np.arange(n)
    np.random.seed(r)
    np.random.shuffle(idx)
    
    df_train = data.iloc[idx[:n_train]].copy()
    df_val = data.iloc[idx[n_train:n_train+n_val]].copy()
    df_test = data.iloc[idx[n_train+n_val:]].copy()
    
    
    df_train = df_train.reset_index(drop=True)
    df_val = df_val.reset_index(drop=True)
    df_test = df_test.reset_index(drop=True)
    
    y_train =  np.log1p(df_train.price.values)
    y_val =  np.log1p(df_val.price.values)
    y_test =  np.log1p(df_test.price.values)
    
    del df_train['price']
    del df_val['price']
    del df_test['price']
    
    X_null_train = df_train.fillna(0).values
    w0_null, w_null = train_linear_regression(X_null_train, y_train)
    
    X_null_val = df_val.fillna(0).values
    y_null_pred_val = w0_null + X_null_val.dot(w_null)
    rmse_val = np.round(rmse(y_val, y_null_pred_val),2)
    
    rmse_list.append(rmse_val)
    
    print(r, w0_null, rmse_val)

0 -427.6284462742615 0.65
1 -414.70554911217323 0.65
2 -423.25413364791126 0.65
3 -421.8933160047171 0.64
4 -422.06099099761855 0.64
5 -429.1787522759954 0.63
6 -427.4918954637153 0.63
7 -428.1884698705238 0.65
8 -427.69291206704054 0.65
9 -426.98268971952757 0.64


In [98]:
rmse_list

[0.65, 0.65, 0.65, 0.64, 0.64, 0.63, 0.63, 0.65, 0.65, 0.64]

In [99]:
np.round(np.std(rmse_list), 3)

0.008

The standard deviation is 0.008

In [104]:
r = 9 

idx = np.arange(n)
np.random.seed(r)
np.random.shuffle(idx)

df_train = data.iloc[idx[:n_train]].copy()
df_val = data.iloc[idx[n_train:n_train+n_val]].copy()
df_test = data.iloc[idx[n_train+n_val:]].copy()

df_train_val = pd.concat([df_train, df_val])

df_train_val = df_train_val.reset_index(drop=True)
df_test = df_test.reset_index(drop=True)

y_train_val =  np.log1p(df_train_val.price.values)
y_test =  np.log1p(df_test.price.values)

del df_train_val['price']
del df_test['price']

In [105]:
X_train_val = df_train_val.fillna(0).values
w_0_train_val, w_train_val = train_linear_regression_reg(X_train_val, y_train_val, r = 0.001)

X_test = df_test.fillna(0).values
y_test_pred = w_0_train_val + X_test.dot(w_train_val)

np.round(rmse(y_test, y_test_pred),2)

0.65